### PR TITLE
BL-1052-bd-design-development-plan

### PR DESCRIPTION
**Pull request by dpearson00**
## Description

Create Design Documentation for the Java Backend. As well as Jira ticket suggestions. 

#### Video Link

[Loom Video](https://www.loom.com/share/941ccc039cd2472c8ddcb621bde39df7)

#### Jira Link

[BL-1052](https://bloomtechlabs.atlassian.net/browse/BL-1052)

## Checklist:

- [x] My design/diagram follows the style guidelines of this project
- [x] My docs are organized within an appropriate folder structure in our repo
- [x] Pull request description clearly describes docs made & motivations for said additions

### CODE REVIEW

**Review by sarahlineb — COMMENTED** [review]
very organized, easy to read, helpful links and information

**Review by razborovdb — DISMISSED** [review]
The database schema, endpoints, and Jira ticket suggestions cover endpoints for creating the java backend application.

File: backend_design_doc.md [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+| [PUT]    | /profile/{profile_id}             | Updates specific profile.                                                                                                |
+
+#### User:
+
+<p>These endpoints are user-focused. As opposed to the more flexible Profile endpoints where <b>profile_id</b> must be specified, these endpoints retrieve data specific only to the user profile that is making the API request by using the logged-in user's <b>profile_id</b>.</p>
+
+| Method | URL              | Description                                                                 |
+| ------ | ---------------- | --------------------------------------------------------------------------- |
+| [GET]  | /user/           | Returns an event object with the specified `type`.                          |
+| [GET]  | /user/inbox/     | Returns an event object                                                     |
```
Comment (deepalisethia) [inline backend_design_doc.md]: Maybe add periods at the end of the sentences on line 33 and 34 to keep with the pattern.

File: backend_design_doc.md [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+  parent_id: INTEGER (required, foreign key),
+}
+```
+
+| Method   | URL                                                | Description                                                                        |
+| -------- | -------------------------------------------------- | ---------------------------------------------------------------------------------- |
+| [GET]    | /children                                          | Returns an array containing all existing children.                                 |
+| [POST]   | /children                                          | Requires a username, name, and age. Returns the name, profile_id, and parent_id.   |
+| [GET]    | /children/{child_id}                               | Returns the child with the given 'id'.                                             |
+| [PUT]    | /children/{child_id}                               | Returns the updated child object                                                   |
```
Comment (deepalisethia) [inline backend_design_doc.md]: periods on 79 and 80 as well

File: backend_design_doc.md [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+  profile_id: INTEGER (required, foreign key),
+  status: STRING (required, default: 'pending'),
+  approved_by: INTEGER (optional, foreign key, default: null, references admin_id),
+}
+```
+
+| Method   | URL                                | Description                                                                                           |
+| -------- | ---------------------------------- | ----------------------------------------------------------------------------------------------------- |
+| [GET]    | /instructor/courses/               | Returns an array containing all course event objects belonging to the currently logged in instructor. |
+| [POST]   | /instructor/register               | Registers an instructor with data                                                                     |
```
Comment (deepalisethia) [inline backend_design_doc.md]: periods on 105, 106

File: backend_design_doc.md [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+  location: STRING (required),
+  number_of_sessions: INTEGER (required),
+  instructor_name: STRING (required),
+  syllabus_link: STRING (required)
+}
+```
+
+| Method   | URL                 | Description                                                                                                                            |
+| -------- | ------------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
+| [GET]    | /course             | Returns an array containing all course objects                                                                                         |
```
Comment (deepalisethia) [inline backend_design_doc.md]: periods on 153, 157

File: backend_design_doc.md [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+  posted_at: TIMESTAMP (auto-generated),
+}
+```
+
+| Method   | URL                      | Description                                                                                   |
+| -------- | ------------------------ | --------------------------------------------------------------------------------------------- |
+| [GET]    | /newsfeed/               | Returns an array containing all newsfeed objects.                                             |
+| [GET]    | /newsfeed/{newsfeed_id}  | Returns the event object with the specified `newsfeed_id`.                                    |
+| [POST]   | /newsfeed/               | Creates a new newsfeed object and returns the newly created newsfeed.                         |
+| [PUT]    | /newsfeed/{newsfeed_id}  | Updates the newsfeed object with the given newsfeed_id and returns the newly updated newsfeed |
```
Comment (deepalisethia) [inline backend_design_doc.md]: period

File: backend_design_doc.md [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+{
+  conversation_id: INCREMENT (primary key, auto-increments, generated by database),
+  profile_id: INTEGER (required, foreign key),
+}
+```
+
+| Method   | URL                           | Description                                                                                                                      |
+| -------- | ----------------------------- | -------------------------------------------------------------------------------------------------------------------------------- |
+| [GET]    | /conversation_id/             | Returns an array filled with inbox event objects.                                                                                |
+| [GET]    | /conversation_id/{profile_id} | Retrieves an inbox with the specified inbox_id                                                                                   |
```
Comment (deepalisethia) [inline backend_design_doc.md]: period

File: backend_design_doc.md [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+  content: STRING (required),
+  details: STRING (required),
+  profile_id: INTEGER (required, foreign key, auto-generated by database),
+}
+```
+
+| Method   | URL                          | Description                                                                         |
+| -------- | ---------------------------- | ----------------------------------------------------------------------------------- |
+| [GET]    | /calendar-events/            | Returns an array filled with calendar event objects.                                |
+| [GET]    | /calendar-events/user/       | Retrieves calendar event objects with the profile_id of logged in user              |
```
Comment (deepalisethia) [inline backend_design_doc.md]: period on 214, 215

**Review by deepalisethia — CHANGES_REQUESTED** [review]
I'm not sure if punctuation matters at all - but since it is a design document, I assume that it does. I would just review where I left comments. If it doesn't matter so much - then please ignore and it looks pretty good to me. Clear, concise and to the point.

**Review by razborovdb — DISMISSED** [review]

**Review by deepalisethia — DISMISSED** [review]
Looks good to me!

**Review by sarahlineb — DISMISSED** [review]

**Review by sarahlineb — APPROVED** [review]
👍

**Review by deepalisethia — APPROVED** [review]
Looks good.

**Review by razborovdb — APPROVED** [review]

**Review by Jwatk13 — COMMENTED** [review]
Great layout and organization of the endpoints and ticket suggestions.

**Comment by enbytedev** [comment]
It looks good! It's a judgement call, especially given how much things can still change, but perhaps consider specifying HTTP response codes, the response body, etc. and the conditions for each. Perhaps lay out the middleware on each route, for example a "must be logged in" middleware may save repetition. This way everybody is on the same page.

**Review by confect1ondev — APPROVED** [review]
It looks good! Perhaps lay out the middleware on each route, for example a "must be logged in" middleware may save repetition. This way everybody is on the same page.

**Comment by tkozzer** [comment]
> It looks good! Perhaps lay out the middleware on each route, for example a "must be logged in" middleware may save repetition. This way everybody is on the same page.

If I'm understanding you correctly, I think this will be solved by the AuthInterceptor. So when somebody tries to hit an endpoint the first step will be to check if they are authenticated with a valid JWT. Then once they pass that stage, they can't proceed to an endpoint until we check that they have the appropriate permissions/roles. So the interceptor will be used to either allow a request to move forward or to return a 403 error.

There is a debate on whether to use a filter or an interceptor for handling this though. More on this debate can be found [here](https://stackoverflow.com/questions/35856454/difference-between-interceptor-and-filter-in-spring-mvc)

**Review by AlexiusThomas — APPROVED** [review]
Code looks dry and good job!

**Pull request merged**